Assets/_BuildingBlocks/Credentials/Scripts/KeyVault.cs

```diff
@@ -10,317 +10,212 @@
-            request.SetRequestHeader("Authorization", "Basic " + Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(username + ":" + password)));
+            var request = UnityWebRequest.Get(queryStringAuthorization.GetFullUri());
+            if (log) Debug.Log("Trying query sting auth: " + queryStringAuthorization.GetFullUri());
             yield return request.SendWebRequest();
-
             if (IsAuthorized(request))
             {
-                if (log) Debug.Log("Access granted with username and password for: " + uri);
-                NewURLAuthorizationDetermined(uri, AuthorizationType.UsernamePassword, username: username, password: password);
-            }
-            else
-            {
-                Debug.LogError("Access denied with username and password for: " + uri);
-                //todo: send an event here so UI can update
+                if (log) Debug.Log("Access granted with authorization type: " + queryStringAuthorization.GetType() + " for: " + queryStringAuthorization.GetFullUri());
+                NewAuthorizationDetermined(queryStringAuthorization);
             }
         }
 
-        private IEnumerator FindSpecificAuthorizationType(Uri uri, string key)
+        private IEnumerator TryAllSupportedAuthorizations(Uri uri, string username, string passwordOrKey)
         {
-            AuthorizationType foundType = AuthorizationType.FailedOrUnsupported;
-
-            //Start with resetting this url history to unkown
-            ClearURLFromStoredAuthorizations(uri);
-
-            // Try a request without credentials
-            var noCredentialsRequest = UnityWebRequest.Get(uri);
-            yield return noCredentialsRequest.SendWebRequest();
-            if (IsAuthorized(noCredentialsRequest))
+            foreach (var supportedAuthorization in supportedAuthorizationTypes.Keys)
             {
-                if (log) Debug.Log("Found no credentials needed for this layer: " + uri);
-                foundType = AuthorizationType.Public;
-                NewURLAuthorizationDetermined(uri, foundType);
-                yield break;
+                yield return TrySupportedAuthorization(supportedAuthorization, uri, username, passwordOrKey);
+                if (storedAuthorizations.ContainsKey(uri)) yield break; // if the Auth test was succesful, stop looking.
             }
-            
-            // Try input key as bearer token
-            var bearerTokenRequest = UnityWebRequest.Get(uri);
-            bearerTokenRequest.SetRequestHeader("Authorization", "Bearer " + key);
-            yield return bearerTokenRequest.SendWebRequest();
-            if (IsAuthorized(bearerTokenRequest))
+        }
+
+        private IEnumerator TrySupportedAuthorization(Type authType, Uri uri, string username, string passwordOrKey)
+        {
+            var potentialAuth = CreateStoredAuthorization(authType, uri, passwordOrKey, username); //passwordOrKey before userName to get the correct constructor order
```

Comment:
Als je je zorgen maakt om volgorde: gebruik named arguments. Dat is misschien wel duidelijker dan deze comment

Reply:
het probleem is hier dat het een params string[] is omdat verschillende credential types een verschillend aantal constructor argumenten heeft (1 voor public, 2 voor key, code, (bearer)token, of 3 voor UsernamePassword. de eerste is altijd de URL. Ik wil 1 generieke functie hebben die ze allemaal kan maken zodat ik kan loopen over die dictionary. In dit geval is voor UsernamePassword de volgorde password, userName, omdat dan het password argument ook gebruikt kan worden voor key/code/token